### PR TITLE
Remove IExcludeTypeFilter

### PR DESCRIPTION
**Pull request by rynowak**
- Removes IExcludeTypeFilter
- Replaced with a property 'ValidateChildren' on ModelMetadata
- Teach ValidationVisitor to respect 'ValidateChildren' for enumerable
  types.

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Core/Internal/MvcCoreMvcOptionsSetup.cs [hunk 1]
```diff
@@ -63,13 +64,11 @@ public static void ConfigureMvc(MvcOptions options, IHttpRequestStreamReaderFact
...
-
-            // Any 'known' types that we bind should be marked as excluded from validation.
-            options.ValidationExcludeFilters.Add(typeof(System.Threading.CancellationToken));
-            options.ValidationExcludeFilters.Add(typeof(IFormFile));
-            options.ValidationExcludeFilters.Add(typeof(IFormCollection));
+            options.ModelMetadataDetailsProviders.Add(new ValidationExcludeFilter(typeof(Type)));
+            options.ModelMetadataDetailsProviders.Add(new ValidationExcludeFilter(typeof(Uri)));
+            options.ModelMetadataDetailsProviders.Add(new ValidationExcludeFilter(typeof(CancellationToken)));
+            options.ModelMetadataDetailsProviders.Add(new ValidationExcludeFilter(typeof(IFormFile)));
+            options.ModelMetadataDetailsProviders.Add(new ValidationExcludeFilter(typeof(IFormCollection)));
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Internal/MvcCoreMvcOptionsSetup.cs]: Talked to @Eilon - he wasn't concerned with the loss of the extension method.

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Metadata/ValidationMetadata.cs [hunk 2]
```diff
@@ -19,6 +19,14 @@ public class ValidationMetadata
         public bool? IsRequired { get; set; }
 
         /// <summary>
+        /// Gets or sets a value that indicates whether children of the model should be validated. If <c>null</c>
+        /// then <see cref="ModelMetadata.ValidateChildren"/> will be computed based on the value of
+        /// <see cref="ModelMetadata.IsComplexType"/> and <see cref="ModelMetadata.IsEnumerableType"/>. By default
+        /// complex types and enumerable types support validation of children.
+        /// </summary>
+        public bool? ValidateChildren { get; set; }
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Metadata/ValidationMetadata.cs]: @dougbu suggestions on how to properly document this?

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs [hunk 3]
```diff
@@ -191,11 +182,16 @@ private bool VisitEnumerableType()
...
+            if (_model != null && _metadata.ValidateChildren)
             {
                 var strategy = _strategy ?? DefaultCollectionValidationStrategy.Instance;
                 isValid = VisitChildren(strategy);
             }
+            else if (_model != null)
+            {
+                // Suppress validation for the prefix, but we still want to validate the object.
+                SuppressValidation(_key);
+            }
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs]: This is the fix for the `JToken` issue.

File: test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/Validation/DefaultObjectValidatorTests.cs [hunk 4]
```diff
@@ -655,13 +656,40 @@ public void Validate_CollectionType_ListOfComplexType_Invalid()
...
+
         [Theory]
-        [InlineData(new[] { "Foo", "Bar", "Baz" }, typeof(string[]))]
-        [InlineData(new[] { 1, 2, 3 }, typeof(int[]))]
-        [InlineData(new[] { "Foo", "Bar", "Baz" }, typeof(IList<string>))]
-        [InlineData(new[] { "Foo", "Bar", "Baz" }, typeof(HashSet<string>))]
-        [InlineData(new[] { "1/1/14", "2/2/14", "3/3/14" }, typeof(ICollection<DateTime>))]
-        [InlineData(new[] { "Foo", "Bar", "Baz" }, typeof(HashSet<Uri>))]
+        [MemberData(nameof(ValidCollectionData))]
         public void Validate_IndexedCollectionTypes_Valid(object model, Type type)
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/Validation/DefaultObjectValidatorTests.cs]: This test was passing by accident. Notice that these are all arrays!

File: test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/Validation/DefaultObjectValidatorTests.cs [hunk 5]
```diff
@@ -734,19 +762,19 @@ public void Validate_CollectionType_DictionaryOfSimpleType_Invalid()
...
             Assert.Empty(entry.Errors);
 
             entry = modelState["items[1].Key"];
-            Assert.Equal(ModelValidationState.Skipped, entry.ValidationState);
+            Assert.Equal(ModelValidationState.Valid, entry.ValidationState);
             Assert.Empty(entry.Errors);
 
             entry = modelState["items[1].Value"];
-            Assert.Equal(ModelValidationState.Skipped, entry.ValidationState);
+            Assert.Equal(ModelValidationState.Valid, entry.ValidationState);
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/Validation/DefaultObjectValidatorTests.cs]: Changes due to the change in whether KVP<,> is considered "simple"

File: test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/Validation/DefaultObjectValidatorTests.cs [hunk 6]
```diff
@@ -890,6 +918,39 @@ public void Validate_ForExcludedType_PropertiesMarkedAsSkipped()
...
+            var validator = CreateValidator(typeof(List<string>));
+
+            var model = new List<string>()
+            {
+                "15",
+            };
+
+            // Note that user.ConfirmPassword has no entry in modelstate - we should not
+            // create one just to mark it as skipped.
+            modelState.SetModelValue("userIds[0]", "15", "15");
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/Validation/DefaultObjectValidatorTests.cs]: Will remove this comment, it's a copy-paste.

File: test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/Validation/DefaultObjectValidatorTests.cs [hunk 7]
```diff
@@ -897,23 +958,20 @@ private static IModelValidatorProvider CreateValidatorProvider()
...
+            var provider = TestModelMetadataProvider.CreateDefaultProvider(excludeFilters.ToArray());
+            return new DefaultObjectValidator(provider);
         }
 
-        private static DefaultObjectValidator CreateValidator(params IExcludeTypeValidationFilter[] excludeFilters)
+        private static DefaultObjectValidator CreateValidator(params IMetadataDetailsProvider[] providers)
         {
-            return new DefaultObjectValidator(excludeFilters, TestModelMetadataProvider.CreateDefaultProvider());
+            var provider = TestModelMetadataProvider.CreateDefaultProvider(providers);
+            return new DefaultObjectValidator(provider);
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/Validation/DefaultObjectValidatorTests.cs]: Some cleanup here to make sure that we're always using the same metadata provider in validation and model binding.

File: test/Microsoft.AspNet.Mvc.IntegrationTests/DictionaryModelBinderIntegrationTest.cs [hunk 8]
```diff
@@ -129,14 +129,12 @@ public class DictionaryModelBinderIntegrationTest
...
-
-            // Key and Value are skipped if they have simple types.
-            Assert.Equal(ModelValidationState.Skipped, entry.ValidationState);
+            Assert.Equal(ModelValidationState.Valid, entry.ValidationState);
 
             entry = Assert.Single(modelState, kvp => kvp.Key == "parameter[low].Value").Value;
             Assert.Equal("10", entry.AttemptedValue);
             Assert.Equal("10", entry.RawValue);
-            Assert.Equal(ModelValidationState.Skipped, entry.ValidationState);
+            Assert.Equal(ModelValidationState.Valid, entry.ValidationState);
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/DictionaryModelBinderIntegrationTest.cs]: Changes due to the change in whether KVP<,> is considered "simple"

File: test/Microsoft.AspNet.Mvc.IntegrationTests/ValidationIntegrationTests.cs [hunk 9]
```diff
@@ -1167,6 +1169,55 @@ private class Country
             Assert.Equal(ModelValidationState.Valid, entry.ValidationState);
         }
 
+        [Fact]
+        public async Task FromBody_JToken_ExcludedFromValidation()
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/ValidationIntegrationTests.cs]: Integration test for #3676

File: test/Microsoft.AspNet.Mvc.Test/MvcOptionsSetupTest.cs [hunk 10]
```diff
@@ -134,7 +137,7 @@ public void Setup_IgnoresAcceptHeaderHavingWildCardMediaAndSubMediaTypes()
         }
 
         [Fact]
-        public void Setup_SetsUpExcludeFromValidationDelegates()
+        public void Setup_SetsUpMetadataDetailsProviders()
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Test/MvcOptionsSetupTest.cs]: I'm really not a fan of this kind of test in general. I updated it, but IMO these are super low value.

File: src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ModelMetadata.cs [hunk 11]
```diff
@@ -290,6 +290,11 @@ public string PropertyName
         public abstract string TemplateHint { get; }
 
         /// <summary>
+        /// Gets a value that indicates whether children of the model should be validated.
```
Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ModelMetadata.cs]: Not sure how often we use the term "child" but maybe clarifying whether it means sub-properties, collection items, etc.?

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs [hunk 12]
```diff
@@ -492,6 +493,31 @@ public override string TemplateHint
...
         /// <inheritdoc />
+        public override bool ValidateChildren
+        {
+            get
+            {
+                if (!_validateChildren.HasValue)
+                {
+                    if (ValidationMetadata.ValidateChildren.HasValue)
+                    {
+                        _validateChildren = ValidationMetadata.ValidateChildren.Value;
```
Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs]: Can you upgrade to +4 Pranav points by doing:

```c#
_validateChildren = ValidationMetadata.ValidateChildren?.Value ?? (IsComplexType || IsEnumerableType);
```

:smile:

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationExcludeFilter.cs [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+        {
+            if (type == null)
+            {
+                throw new ArgumentNullException(nameof(type));
+            }
+
+            Type = type;
+        }
+
+        public ValidationExcludeFilter(string typeFullName)
```
Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationExcludeFilter.cs]: `fullTypeName`? (Including renaming other properties, etc.)

File: src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ModelMetadata.cs (hunk 11 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ModelMetadata.cs]: sure bro

**Comment by Eilon** [comment]
:shipit: from me.

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Metadata/ValidationMetadata.cs (hunk 2 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Metadata/ValidationMetadata.cs]: Reads fairly well though the last sentence switches perspectives. Suggest: `... and <see cref="ModelMetadata.IsEnumerableType"/>: <see cref="ModelMetadata.ValidateChildren"/> will be <c>true</c> if either <see cref="ModelMetadata.IsComplexType"/> or <see cref="ModelMetadata.IsEnumerableType"/> is <c>true</c>; <c>false</c> otherwise.`

Hmm, that's a bit wordy. Follow your :heart:

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationExcludeFilter.cs [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+using System;
+using System.Diagnostics;
+using System.Reflection;
+using Microsoft.AspNet.Mvc.ModelBinding.Metadata;
+
+namespace Microsoft.AspNet.Mvc.ModelBinding.Validation
+{
+    public class ValidationExcludeFilter : IValidationMetadataProvider
+    {
+        public ValidationExcludeFilter(Type type)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationExcludeFilter.cs]: `+ ///` throughout this new, visible class.

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationExcludeFilter.cs [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+            }
+
+            TypeFullName = typeFullName;
+        }
+
+        public Type Type { get; }
+
+        public string TypeFullName { get; }
+
+        public void GetValidationMetadata(ValidationMetadataProviderContext context)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationExcludeFilter.cs]: Are we assuming the composite provider has already `null`-checked `context`? If not need a check here.

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationExcludeFilter.cs [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+            if (TypeFullName != null)
+            {
+                if (IsMatchingName(context.Key.ModelType))
+                {
+                    context.ValidationMetadata.ValidateChildren = false;
+                }
+
+                return;
+            }
+        }
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationExcludeFilter.cs]: `+` a `Debug.Assert(false);` if we reach here.

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationExcludeFilter.cs [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+
+        private bool IsMatchingName(Type type)
+        {
+            Debug.Assert(TypeFullName != null);
+
+            if (type == null)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationExcludeFilter.cs]: How can `context.Key.ModelType` be `null`?

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationExcludeFilter.cs [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+                {
+                    context.ValidationMetadata.ValidateChildren = false;
+                }
+
+                return;
+            }
+
+            if (TypeFullName != null)
+            {
+                if (IsMatchingName(context.Key.ModelType))
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationExcludeFilter.cs]: Need to document somewhere that interface implementations can be excluded using the `ValidationExcludeFilter(Type)` constructor but not using the `ValidationExcludeFilter(string)` constructor.

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationExcludeFilter.cs (hunk 17 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationExcludeFilter.cs]: This is called recursively. `context.Key.ModelType` won't be null, but `context.Key.ModelType.BaseType` might be

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs [hunk 19]
```diff
@@ -191,11 +182,16 @@ private bool VisitEnumerableType()
...
 
-            if (_model != null)
+            if (_model != null && _metadata.ValidateChildren)
             {
                 var strategy = _strategy ?? DefaultCollectionValidationStrategy.Instance;
                 isValid = VisitChildren(strategy);
             }
+            else if (_model != null)
+            {
+                // Suppress validation for the prefix, but we still want to validate the object.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs]: Won't `SuppressValidation()` also set the model's entry to `Skipped`? Put another way, is the comment intended to mean the exact match for `_key` will be updated a few lines later?

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs]: If it does set the model's entry to skipped, then the subsequent code sets it back to valid. The end result is that the node for `_model` gets validated.

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/SimpleTypesExcludeFilter.cs [hunk 20]
```diff
@@ -1,64 +0,0 @@
...
-        protected virtual bool IsSimpleType(Type type)
-        {
-            return type.GetTypeInfo().IsPrimitive ||
-                type.Equals(typeof(decimal)) ||
-                type.Equals(typeof(string)) ||
-                type.Equals(typeof(DateTime)) ||
-                type.Equals(typeof(Guid)) ||
-                type.Equals(typeof(DateTimeOffset)) ||
-                type.Equals(typeof(TimeSpan)) ||
-                type.Equals(typeof(Uri));
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/SimpleTypesExcludeFilter.cs]: Where's the code that adds `ValidationExcludeFilter`s for these simple types?

Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/SimpleTypesExcludeFilter.cs]: Never mind. It's in `DefaultModelMetadata` and removes this alternate definition of a "simple type".

File: test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/Validation/DefaultObjectValidatorTests.cs [hunk 21]
```diff
@@ -5,6 +5,7 @@
 using System.Collections.Generic;
 using System.ComponentModel.DataAnnotations;
 using System.Linq;
+using Microsoft.AspNet.Mvc.ModelBinding.Metadata;
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/Validation/DefaultObjectValidatorTests.cs]: Sortification required

File: test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/Validation/DefaultObjectValidatorTests.cs (hunk 4 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/Validation/DefaultObjectValidatorTests.cs]: What was important was the `type` parameter, not the runtime type of `model`. But, sure, can use instances matching the checked `type`.

File: test/Microsoft.AspNet.Mvc.IntegrationTests/DictionaryModelBinderIntegrationTest.cs (hunk 8 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/DictionaryModelBinderIntegrationTest.cs]: It's coming back to me: A collection of simple types and a `KVP<,>` of simple types cannot have any applied validations. It's just useless to validate them. Excluding `KVP<,>` made perfect sense though I'm not sure why `IList<string>` wasn't automatically excluded as well.

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs (hunk 19 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs]: Think the comment would be more helpful if it were `Suppress validation for the entries. May set the model's entry to Skipped but subsequent code will validate that collection.`

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs (hunk 12 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadata.cs]: :trollface:

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Metadata/ValidationMetadata.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Metadata/ValidationMetadata.cs]: I like yours better.

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationExcludeFilter.cs (hunk 15 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationExcludeFilter.cs]: :+1:

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationExcludeFilter.cs (hunk 18 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/ValidationExcludeFilter.cs]: Added to constructor docs.

**Comment by rynowak** [comment]
@dougbu updated

File: test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/Validation/DefaultObjectValidatorTests.cs (hunk 6 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/ModelBinding/Validation/DefaultObjectValidatorTests.cs]: Your :up::date: left this important, totally rad improvement out. :disappointed:

**Comment by dougbu** [comment]
:shipit:
